systems/plants/@RigidBodyManipulator/bodyKin.m

```diff
@@ -13,89 +13,151 @@
+      dT = kinsol.dTdq{body_ind}*Tframe;
+      ddT = kinsol.ddTdqdq{body_ind}*Tframe;
+      [invT, dinvT, ddinvT] = invHT(T,dT,ddT);
+    elseif compute_first_derivatives
+      dT = kinsol.dTdq{body_ind}*Tframe;
+      [invT, dinvT] = invHT(T,dT);
+    else
+      invT = invHT(T);
+    end
+    x = invT*pts;
+    x = x(1:3,:);
+    
     if compute_P
-      dP = zeros(3*m,3*m,nq);
-      dinvT_reshaped = permute(reshape(dinvT,[nq,3,4]),[2,3,1]);
+      P = zeros(3*m,3*m)*kinsol.q(1); % keep the taylorvars happy
+      for i=1:size(pts,2)
+        P((i-1)*3+1:i*3,(i-1)*3+1:i*3)=invT(1:3,1:3);
+      end
     end
-    for i=1:m
-      grad = dinvT*pts(:,i);
-      J((i-1)*3+(1:3),:) = reshape(grad,nq,3)';
+    if compute_first_derivatives
+      nq = size(kinsol.q,1);
+      J = zeros(3*m,nq);
       if compute_P
-        dP((i-1)*3+1:i*3,(i-1)*3+1:i*3,:)=dinvT_reshaped(1:3,1:3,:);
+        dP = zeros(3*m,3*m,nq);
+        dinvT_reshaped = permute(reshape(dinvT,[nq,3,4]),[2,3,1]);
+      end
+      for i=1:m
+        grad = dinvT*pts(:,i);
+        J((i-1)*3+(1:3),:) = reshape(grad,nq,3)';
+        if compute_P
+          dP((i-1)*3+1:i*3,(i-1)*3+1:i*3,:)=dinvT_reshaped(1:3,1:3,:);
+        end
       end
+      dP = reshape(dP,(3*m)^2,nq);
     end
-    dP = reshape(dP,(3*m)^2,nq);
-  end
-  if compute_second_derivatives
-    if isempty(kinsol.ddTdqdq{body_ind})
-      error('you must call doKinematics with the second derivative option enabled');
+    if compute_second_derivatives
+      if isempty(kinsol.ddTdqdq{body_ind})
+        error('you must call doKinematics with the second derivative option enabled');
+      end
+      dJ_reshaped = ddinvT*pts;
+      dJ = reshape(permute(reshape(dJ_reshaped,[nq,3,nq,m]),[2,4,3,1]),3*m,nq^2);
     end
-    dJ_reshaped = ddinvT*pts;
-    dJ = reshape(permute(reshape(dJ_reshaped,[nq,3,nq,m]),[2,4,3,1]),3*m,nq^2);
+    
   end
-  
 end
 
 
 end
+
+function [x,P,J,dP,dJ] = bodyKinNew(obj,kinsol,body_or_frame_ind,pts)
+compute_P = nargout > 1;
+compute_J = nargout > 2;
+compute_dP = nargout > 3;
+compute_dJ = nargout > 4;
+
+if compute_dJ
+  [x, J, dJ] = forwardKin(obj, kinsol, 1, pts, 0, body_or_frame_ind);
+elseif compute_J
+  [x, J] = forwardKin(obj, kinsol, 1, pts, 0, body_or_frame_ind);
+else
+  x = forwardKin(obj, kinsol, 1, pts, 0, body_or_frame_ind);
+end
+
+m = size(pts,2);
+if (kinsol.mex)
+  if (obj.mex_model_ptr==0)
+    error('Drake:RigidBodyManipulator:InvalidKinematics','This kinsol is no longer valid because the mex model ptr has been deleted.');
+  end
+  if ~isnumeric(pts)
+    error('Drake:RigidBodyManipulator:InvalidKinematics','This kinsol is not valid because it was computed via mex, and you are now asking for an evaluation with non-numeric pts.  If you intended to use something like TaylorVar, then you must call doKinematics with use_mex = false');
+  end
+  if compute_dP
+    [P, dP] = forwardKinPositionGradientmex(obj.mex_model_ptr, m, body_or_frame_ind, 1);
```

Comment:
Decided to put these in their own mex file. I really don't like the output of this function, with the alternating q and point gradients.